web/text_accessibility.js

```diff
@@ -0,0 +1,250 @@
+/* Copyright 2022 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { binarySearchFirstItem, shadow } from "pdfjs-lib";
+
+/**
+ * This class aims to provide some methods:
+ *  - to reorder elements in the DOM with respect to the visual order;
+ *  - to create a link, using aria-owns, between spans in the textLayer and
+ *    annotations in the annotationLayer. The goal is to help to know
+ *    where the annotations are in the text flow.
+ */
+class TextAccessibilityManager {
+  #textLayer;
+
+  #textChildren;
+
+  #textNodes = new Map();
+
+  #waitingElements = new Set();
+
+  #pageDiv;
+
+  /**
+   * @param {HTMLDivElement} pageDiv
+   */
+  constructor(pageDiv) {
+    this.#pageDiv = pageDiv;
+  }
+
+  get #hasTextLayer() {
+    return !!this.#pageDiv.querySelector(":scope > .textLayer .endOfContent");
+  }
+
+  get _textLayerElements() {
+    // When zooming the text layer is removed from the DOM and sometimes
+    // it's rebuilt hence the nodes are no longer valid.
+
+    const textLayer = this.#pageDiv.getElementsByClassName("textLayer").item(0);
```

Comment:
Similar to the `TextHighlighter`-class, why can't you add a `setTextMapping`-method in this class instead and call that one using something like the following
```javascript
this.accessibilityManager?.setTextMapping(this.textDivs);
```
placed just *after* [this line](https://github.com/mozilla/pdf.js/blob/e1e97c6edd8729e968bde47f60a1901e8e649901/web/text_layer_builder.js#L99)?

That way, the DOM elements should be directly available without the need for any additional DOM lookups here.